pkg/elasticetcd/leader.go

```diff
@@ -0,0 +1,269 @@
+		}
+	}
+
+	// Update the nominee list after the nominee removals
+	nominees = diffStringSlices(nominees, unvolunteered)
+	nomineeCount := len(nominees)
+	ee.log.WithField("nominees", nominees).Debug("updated nominees list")
+
+	// Prepare a map of volunteers names and their published CURLs
+	volunteersMap, err := urlsMapFromGetResp(volunteersResp, volunteerPrefix)
+	if err != nil {
+		ee.log.WithError(err).Error("could not prepare volunteers map")
+		return
+	}
+
+	switch {
+	// If idealSize is not met, nominate more servers till the size is met
+	case nomineeCount < ee.conf.IdealSize:
+		// You cannot do nominations if all volunteers have been nominated
+		if compareStringSlices(nominees, volunteers) {
+			ee.log.Debug("all available volunteers have been nominated")
+			return
+		}
+
+		// Filter out already nominated servers
+		available := diffStringSlices(volunteers, nominees)
+
+		// Keep nominating in a round-robin fashion till the required nominations are done
+		for _, h := range available {
+			err := ee.nominate(h, volunteersMap[h])
+			if err != nil {
+				ee.log.WithError(err).WithField("host", h).Error("failed to nominate host")
+				continue
+			}
+			ee.log.WithField("host", h).Debug("nominated new host")
+			nomineeCount++
+			if nomineeCount == ee.conf.IdealSize {
+				break
+			}
+		}
+
+		// If idealSize is exceeded, remove server nominations till idealSize is reached
+	case nomineeCount > ee.conf.IdealSize:
+		// Remove nominations in a round-robin fashion till the required nominations are removed
+		for _, h := range nominees {
+			if h == ee.conf.Name {
+				// skip yourself
+				continue
+			}
+			if err := ee.removeNomination(h); err != nil {
+				ee.log.WithError(err).WithField("host", h).Warn("could not remove nomination for host")
+				continue
+			}
+			nomineeCount--
+			if nomineeCount == ee.conf.IdealSize {
+				break
+			}
+		}
+	}
+
+	ee.log.WithField("nomineecount", nomineeCount).Debug("finished doing nominations")
+}
+
+func (ee *ElasticEtcd) nominate(host string, urls types.URLs) error {
+	logger := ee.log.WithField("host", host)
+	logger.Debug("nominating host")
+
+	// Add the new host to the nominees list.
+	// Wait a little for the nominee to pick up the nomination.
+	// Then add the new host as an etcd member.
+	// If we add the new host as a member first, the embeded etcd will begin
+	// trying to connect to the newly added member and which causes new requests
+	// to be blocked, for example the PUT request to add the new nominee.
+	// If we don't wait for a little while before adding the nominee as a member,
+	// the nominee will not be able to read the nominee list it requires to start
+	// its server.
+	// This is only required for the first nominee added. When more than one
+	// server is present the etcd cluster will serve requests when a member is
+	// added.
+	// TODO: Try to figure a way to avoid the sleeping
+
+	ee.addToNominees(host, urls)
+
+	time.Sleep(2 * time.Second)
+
```

Comment:
The sleep here kinda takes care of this as a side effect most of the times. Maybe we just leave it here forever.